service/history/workflow/update/update.go

```diff
@@ -169,100 +162,70 @@ func newCompleted(
+		if ctx.Err() == nil && stCtx.Err() == nil {
+			return nil, err
+		}
 
-// WaitOutcome observes this Update's completion, returning when the Outcome is
-// available. This call will block until the Outcome is known or the provided
-// context.Context expires. It is safe to call this method outside of workflow
-// lock.
-func (u *Update) WaitOutcome(ctx context.Context) (UpdateStatus, error) {
-	outcome, err := u.outcome.Get(ctx)
-	if err != nil {
-		return UpdateStatus{enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, outcome}, err
+		if ctx.Err() != nil {
+			// Handle root context deadline expiry as normal error which is returned to the caller.
+			return nil, ctx.Err()
+		}
 	}
-	return UpdateStatus{enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome}, nil
-}
 
-// WaitAccepted blocks on the acceptance of this update, returning nil if has
-// been accepted but not yet completed or the overall Outcome if the update has
-// been completed (including completed by rejection). This call will block until
-// the acceptance occurs or the provided context.Context expires.
-// It is safe to call this method outside of workflow lock.
-func (u *Update) WaitAccepted(ctx context.Context) (UpdateStatus, error) {
-	if u.outcome.Ready() {
-		// Being complete implies being accepted; return the completed outcome
-		// here because we can.
-		return u.WaitOutcome(ctx)
-	}
-	fail, err := u.accepted.Get(ctx)
-	if err != nil {
-		return UpdateStatus{enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil}, err
-	}
-	if fail != nil {
-		outcome := &updatepb.Outcome{
-			Value: &updatepb.Outcome_Failure{Failure: fail},
+	// Update is not completed but maybe it is accepted.
+	if u.accepted.Ready() || waitStage == enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED {
+		// Using same context which might be already expired, but if accepted future is ready
+		// then it will return immediately without checking context deadline.
+		rejection, err := u.accepted.Get(stCtx)
+		if err == nil {
+			if rejection != nil {
+				return statusRejected(rejection), nil
+			}
+			return statusAccepted(), nil
+		}
+
+		// See comment for the same check in "completed" branch.
+		if ctx.Err() == nil && stCtx.Err() == nil {
+			return nil, err
+		}
+
+		if ctx.Err() != nil {
+			// Handle root context deadline expiry as normal error which is returned to the caller.
+			return nil, ctx.Err()
 		}
-		return UpdateStatus{enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome}, nil
 	}
-	return UpdateStatus{enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED, nil}, nil
+
+	// Only get here:
+	//   if stCtx.Err() is not nil, which means that soft timeout (but not user timeout)
+	//     waiting for ACCEPTED/COMPLETED has expired,
+	//   or
+	//   it waitStage is UNSPECIFIED or ADMITTED,
+	// In both cases behavior is the same: return ADMITTED (as the most reached state) and empty outcome.
+	return statusAdmitted(), nil
```

Comment:
I believe this function can be called from Update Workflow and Poll Update. This changes makes perfect sense for Poll Update where the user doesn't specify a wait stage and receives the latest one. For Update Workflow, UNSPECIFIED is turned into COMPLETED, which is fine. But I don't see anything stopping the user from requesting "ADMITTED" now. Is that true; and if so, intendend?

Reply:
Yes, I am getting prepared to support `ADMITTED` as wait stage. It is validated (not allowed) on the caller side but I want all other code support it.